### PR TITLE
[JENKINS-50181] Test and updated dependencies for changes in ssh-credentials for keys without a newline

### PR DESCRIPTION
**Pull request by warden**
[JENKINS-50181](https://issues.jenkins-ci.org/browse/JENKINS-50181)

### CODE REVIEW

**Comment by warden** [comment]
There you go! I see the build is failing because of _spotbugs_ checks, but the tests are passing fine... and I didn't touch those files in scope of this change :)

File: pom.xml
Comment (jvz) [inline pom.xml]:
```suggestion
      <version>1.17.3</version>
```

Freshly released!

**Comment by jvz** [comment]
As for not touching things, the dependencies are things ;)

**Comment by warden** [comment]
Thanks! Feel free to squash merge :)

**Comment by jvz** [comment]
Looks like you've got some compiler warnings preventing this from building successfully. Can you take a look?

**Comment by warden** [comment]
I fixed the spotbugs warnings but I'm not 100% sure that I did it correctly, as my change was totally irrelevant to this (probably with the parent plugin bump, new spotbugs checks were enabled).
@jglick can you double check if my assumptions were correct here?

**Comment by warden** [comment]
@jglick, sorry I wasn't unclear, I know you're not maintaining it :) I just blamed the lines that were causing the spotbugs warnings and I thought that you might have some good tips on the best approach to fix it (or just confirm that my assumptions are fine)

**Comment by warden** [comment]
While doing this simple fix and with the hassle of testing it, I'm just wondering if there is a real benefit of keeping the `ssh-credentials` plugin separate instead of merging it into here?
It seems that https://plugins.jenkins.io/ssh-credentials has no other downward dependencies.
@jvz, @stephenc WDYT?

**Comment by jvz** [comment]
Good question. For me personally, I only have merge access for ssh-credentials, not ssh-agent. There might be some proprietary Jenkins plugins that use it, though.

**Comment by stephenc** [comment]
I thought some of the SCM plugins depend directly on ssh-credentials. Also I wonder if it is actually worth the effort... for safety you’d actually end up moving this plugin there rather than that plugin here

**Comment by warden** [comment]
You're right, doing the other way round ( ssh-agent -> ssh-credentials ) makes more sense.
So, let's say for the sake of curiosity, is there any easy way to check what other Jenkins plugins might depend on `ssh-agent-plugin` ?

**Comment by stephenc** [comment]
I just don't see a compelling need to go through the pain of changing.

Keep in mind that you have all sorts of fun failure modes if you try the move:

_Move ssh-agent into ssh-credentials_:

* User has upgraded `ssh-credentials` but not `ssh-agent` so sees two versions of the classes from `ssh-agent` and hilarity ensues

_Move ssh-credentials into ssh-agent_:

* User has upgraded `ssh-credentials` not not `ssh-agent` so doesn't have any versions of the classes from `ssh-agent` and hilarity ensues
* User doesn't have `ssh-agent` installed but has a 3rd party plugin that uses `ssh-credentials` and hilarity ensues.

IMHO it is a very tricky migration to make on a very popular plugin. Lots of pit holes besides the obvious ones I listed above

**Comment by warden** [comment]
Yeah, that might be tricky indeed, but on the other hand I think it doesn't mean we should be avoiding cleaning up our sandbox for the work to be easier :-)
That's the only reason I started this discussion (maybe it's wrong that I did that here and not on the dev mailing list...) to see if you think it's worthwhile from maintainers perspective (as it was pretty painful from debugging perspective for me)

**Comment by warden** [comment]
@jvz any more changes needed from my side here or is this just waiting for the next release cycle to be merged?

**Comment by jvz** [comment]
@warden I'm not sure. I don't have merge or release permissions in this repo; I'm just a maintainer for ssh-credentials.

**Comment by jvz** [comment]
@batmat could you merge this?

File: pom.xml
Comment (jglick) [inline pom.xml]: Note to self: careful! Needs to be tested manually, not just using `JenkinsRule`, due to `pluginFirstClassLoader`.

File: pom.xml
Comment (jglick) [inline pom.xml]: Probably time for the [BOM](https://github.com/jenkinsci/bom#usage).

**Comment by warden** [comment]
@batmat @jglick so what are your thoughts here, does it make sense for me to spend time re-basing and re-pushing it so that it is quickly reviewed and merged?

File: pom.xml
Comment (jglick) [reply pom.xml]:
```console
$ mvn -f jenkinsci/jenkins/war dependency:tree | fgrep sshd
[INFO] +- org.jenkins-ci.modules:sshd:jar:2.6:compile
[INFO] |  +- org.apache.sshd:sshd-core:jar:1.7.0:compile
```

Comment (jglick) [reply pom.xml]: or as of 2.138.4

```
[INFO] +- org.jenkins-ci.modules:sshd:jar:2.4:compile
[INFO] |  \- org.apache.sshd:sshd-core:jar:1.7.0:compile
```

**Comment by statsperformjanos** [comment]
Hi, any update on this ticket? This bug is kinda blocker for me, the SSH key is in AWS Secrets Manager, which strips the last newline off a plain text secret. :(

**Comment by statsperformjanos** [comment]
I added a `#` to the last line which is not stripped by the AWS Secrets Manager, but stripped by `ssh-agent`, so after all, only the desired newline remains in the last line. 😅

**Comment by aries1980** [comment]
Hi, any update on this ticket? This bug is kinda blocker for me, the SSH key is in *AWS Secrets Manager*, which strips the last newline off a plain text secret. :(

I added a `#` to the last line which is not stripped by the *AWS Secrets Manager*, but stripped by `ssh-agent`, so after all, only the desired newline remains in the last line. 😅

File: pom.xml
Comment (jglick) [reply pom.xml:89]: see also #36

**Comment by warden** [comment]
I'm waiting for the maintainer to speak up what should be done here before resolving the conflicts.

**Comment by warden** [comment]
@MRamonLeon can you take a look if it is worth for me rebasing it and it will get merged then?

**Comment by MRamonLeon** [comment]
> @MRamonLeon can you take a look if it is worth for me rebasing it and it will get merged then?

Better to merge with master as Jesse suggested

**Comment by warden** [comment]
> > @MRamonLeon can you take a look if it is worth for me rebasing it and it will get merged then?
> 
> Better to merge with master as Jesse suggested

@MRamonLeon there you go!

File: pom.xml
Comment (warden) [reply pom.xml:111]: https://github.com/jenkinsci/ssh-agent-plugin/pull/37

**Comment by warden** [comment]
C'mon guys, this PR is waiting for more than 12 months now.. 
I understand that a single person may lack time but [surely out of 7 people that were added recently](https://github.com/jenkins-infra/repository-permissions-updater/blame/master/permissions/plugin-ssh-agent.yml), someone could have a look here...
@MRamonLeon @rsandell @batmat

**Review by jvz — APPROVED** [review]
LGTM again.

**Comment by warden** [comment]
@MRamonLeon @rsandell @batmat ping again 🙏

File: src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgentFactory.java
Comment (MRamonLeon) [inline src/main/java/com/cloudbees/jenkins/plugins/sshagent/mina/MinaRemoteAgentFactory.java:81]:
```suggestion
            throw new IllegalStateException("RemoteLauncher has been initialized with null launcher provider. It should not happen");
```

**Review by MRamonLeon — COMMENTED** [review]
Hi @warden sorry for the late response. I've left some comments. It looks great overall but I believe we can improve a bit the spotbugs warnings. Take a look please. I promise to be reactive this time.

File: src/main/java/com/cloudbees/jenkins/plugins/sshagent/jna/JNRRemoteAgentFactory.java
Comment (MRamonLeon) [reply src/main/java/com/cloudbees/jenkins/plugins/sshagent/jna/JNRRemoteAgentFactory.java:56]: Is SpotBugs still complaining @warden or we can remove all the `SuppressFBWarnings`?

**Comment by MRamonLeon** [comment]
I believe we can we can remove all the SuppressFBWarnings annotations after the null checks added @warden

**Comment by warden** [comment]
@MRamonLeon done, let me know if you need a rebase, otherwise feel free to squash-merge.

**Review by MRamonLeon — APPROVED** [review]

**Comment by MRamonLeon** [comment]
I plan to release it on Monday. Thank you for your contribution @warden, very appreciated!

**Pull request merged**
